clang/lib/CodeGen/TargetInfo.cpp

```diff
@@ -7656,6 +7656,52 @@ MipsABIInfo::returnAggregateInRegs(QualType RetTy, uint64_t Size) const {
   return HandleAggregates(RetTy, Size);
 }
 
+static bool mipsCanReturnDirect(const ASTContext& Ctx, const RecordDecl *RD, unsigned& NumCaps, unsigned& NumInts);
+static bool mipsCanReturnDirect(const ASTContext& Ctx, QualType Ty, unsigned& NumCaps, unsigned& NumInts) {
+  // Treat an enum type as its underlying type.
+  if (const EnumType *EnumTy = Ty->getAs<EnumType>())
+    Ty = EnumTy->getDecl()->getIntegerType();
+  if (Ty->isCHERICapabilityType(Ctx)) {
+    NumCaps++;
+  } else if (Ty->isIntegerType()) {
+    NumInts++;
+  } else if (const RecordType *RT = Ty->getAs<RecordType>()) {
+    if (!mipsCanReturnDirect(Ctx, RT->getDecl(), NumCaps, NumInts))
+      return false;
+  } else if (Ty->isConstantArrayType()) {
+    auto CAT = cast<ConstantArrayType>(Ty->getAsArrayTypeUnsafe());
+    if (CAT->getSize().ugt(2))
+      return false;
+    unsigned ArrayCaps = 0;
+    unsigned ArrayInts = 0;
+    if (!mipsCanReturnDirect(Ctx, CAT->getElementType(), ArrayCaps, ArrayInts))
```

Comment:
This branch leaves `NumCaps` and `NumInt` as their initial values, so I think:
```
struct Foo {
    void * __capability x[2];
    int y;
}
```
would erroneously be returned in registers (or any struct that has any number of <=2-element int/cap arrays, and up to 2 other int/cap fields).

Reply:
Indeed! I meant to add the array values but somehow didn't include that...

Reply:
Is there a reason you don't just pass NumCaps/NumInt directly as you do for records?..

Reply:
I meant to multiply it with the number of elements since it could be one or two.